src/ModelContextProtocol/Server/McpServer.cs

```diff
@@ -434,20 +444,48 @@ await originalListToolsHandler(request, cancellationToken).ConfigureAwait(false)
 
     private void SetSetLoggingLevelHandler(McpServerOptions options)
     {
-        if (options.Capabilities?.Logging is not { } loggingCapability)
-        {
-            return;
-        }
-
-        if (loggingCapability.SetLoggingLevelHandler is not { } setLoggingLevelHandler)
-        {
-            throw new McpException("Logging capability was enabled, but SetLoggingLevelHandler was not specified.");
-        }
+        // We don't require that the handler be provided, as we always store the provided
+        // log level to the server.
+        var setLoggingLevelHandler = options.Capabilities?.Logging?.SetLoggingLevelHandler;
 
         RequestHandlers.Set(
             RequestMethods.LoggingSetLevel,
-            (request, cancellationToken) => setLoggingLevelHandler(new(this, request), cancellationToken),
+            (request, cancellationToken) =>
+            {
+                // Store the provided level.
+                if (request is not null)
+                {
+                    if (_loggingLevel is null)
+                    {
+                        Interlocked.CompareExchange(ref _loggingLevel, new(request.Level), null);
+                    }
+
+                    _loggingLevel.Value = request.Level;
+                }
+
+                // If a handler was provided, now delegate to it.
+                if (setLoggingLevelHandler is not null)
+                {
+                    return setLoggingLevelHandler(new(this, request), cancellationToken);
+                }
+
+                // Otherwise, consider it handled.
+                return EmptyResult.CompletedTask;
+            },
             McpJsonUtilities.JsonContext.Default.SetLevelRequestParams,
             McpJsonUtilities.JsonContext.Default.EmptyResult);
     }
+
+    /// <summary>Maps a <see cref="LogLevel"/> to a <see cref="LoggingLevel"/>.</summary>
+    internal static LoggingLevel ToLoggingLevel(LogLevel level) =>
+        level switch
+        {
+            LogLevel.Trace => Protocol.Types.LoggingLevel.Debug,
+            LogLevel.Debug => Protocol.Types.LoggingLevel.Debug,
+            LogLevel.Warning => Protocol.Types.LoggingLevel.Warning,
+            LogLevel.Error => Protocol.Types.LoggingLevel.Error,
+            LogLevel.Critical => Protocol.Types.LoggingLevel.Critical,
+            LogLevel.Information => Protocol.Types.LoggingLevel.Info,
```

Comment:
```suggestion
            LogLevel.Trace => Protocol.Types.LoggingLevel.Debug,
            LogLevel.Debug => Protocol.Types.LoggingLevel.Debug,
            LogLevel.Information => Protocol.Types.LoggingLevel.Info,
            LogLevel.Warning => Protocol.Types.LoggingLevel.Warning,
            LogLevel.Error => Protocol.Types.LoggingLevel.Error,
            LogLevel.Critical => Protocol.Types.LoggingLevel.Critical,
            LogLevel.None => Protocol.Types.LoggingLevel.Emergency
```

Nit: I mostly like keeping the LogLevel's in order. I also think that LogLevel.None is probably closer to LoggingLevel.Emergency than LoggingLevel.Debug although neither is perfect.